Run tests against sqlite and postgres

shouichi opened the pull request:
This is the first step to tackle https://github.com/Shopify/maintenance_tasks/issues/462.

shouichi commented on test/lib/generators/maintenance_tasks/install_generator_test.rb:
```diff
@@ -9,13 +9,20 @@ class InstallGeneratorTest < Rails::Generators::TestCase
+    original_adapter = ENV["DB"]
+
     setup do
+      # Force the sample app to use sqlite. This is because both the dummy app
+      # and the sample use the same postgres database, ended up migration
+      # failure.
+      ENV["DB"] = "sqlite"
```
This is the only tricky part that I failed to find a cleaner solution.

shouichi reviewed (commented):
This PR essentially switches database adapters on database.yml and others are just adjustments except for the part I made a comment.

adrianna-chang-shopify commented:
Hey @Shopify/rails, could I get thoughts / feedback on doing the above ^ to handle testing the install generator properly with PostgreSQL?

shouichi commented on test/dummy/config/database.yml:
```diff
@@ -1,9 +1,6 @@
-#   gem 'sqlite3'
-#
+<% adapter = ENV.fetch("DB") { "sqlite" } %>
+<% case adapter
```
Thanks for the suggestion. The `adapter` variable is used for raising an error when it's unknown, i.e.,

```
<% else %>
	<% raise "unknown adapter: #{adapter"} %>
<% end %>
```

I believe it's a good practice to raise an unknown adapter. Do you still want me to change?

adrianna-chang-shopify replied:
You can still store it in a variable (`<% case (adapter = ENV.fetch("DB")) %>`), but I agree with Gannon that it's more straightforward to set the default in the test helper instead.

adrianna-chang-shopify replied:
Let's still set this default in the `test_helper` 😄

shouichi commented on test/dummy/config/sample_app_postgres_database.yml:
```diff
@@ -0,0 +1,14 @@
+# Database settings for postgres must be in sync with
```
It turns out that rails 6.0 does not allow us to put yaml files in the fixtures directory https://github.com/rails/rails/pull/36884. It's a PR that @gmcgibbon sent, what a coincidence!

shouichi reviewed (commented):
Hi, guys. Thanks for the help! I applied most of the suggestions and it __mostly__ works.

The last thing that needs to be fixed is `MaintenanceTasks::RunTest#test_#reload_status_does_not_use_query_cache` which probabilistically fails when postgres.

shouichi commented:
Found the cause of `MaintenanceTasks::RunTest#test_#reload_status_does_not_use_query_cache` random failure.

When the test is executed first, active record first executes `SHOW search_path` to find out which postgres schemas to search. The solution I came up with is to run a query (e.g., `Run.first`) in setup and it worked. Is there any better solution?

adrianna-chang-shopify commented:
@shouichi you can filter that query out from the `count_uncached_queries` method:
```ruby
    def count_uncached_queries(&block)
      count = 0

      query_cb = ->(*, payload) do
        unless payload[:cached] || payload[:sql].match?("SHOW search_path")
          count += 1
        end
      end
      ActiveSupport::Notifications.subscribed(query_cb,
        "sql.active_record",
        &block)

      count
    end
```

adrianna-chang-shopify commented on .github/workflows/ci.yml:
```diff
@@ -5,13 +5,28 @@ on: [push, pull_request]
-    name: Ruby ${{ matrix.ruby }} / ${{ matrix.gemfile }}
+    name: Ruby ${{ matrix.ruby }} / ${{ matrix.gemfile }} / ${{ matrix.database }}
     strategy:
+      fail-fast: false
```
Why this change?

shouichi commented:
Fixed all the problems, please take a look. Feel free to ping me if I need to rebase or squash.

adrianna-chang-shopify commented on test/dummy/config/database.yml:
```diff
@@ -13,13 +10,36 @@ development:
   <<: *default
   database: db/production.sqlite3
+<% when "postgres" %>
+# Database settings for postgres must be in sync with
+# test/dummy/config/sample_app_postgres_database.yml.
```
Let's ✂️  this comment

adrianna-chang-shopify commented on test/dummy/config/sample_app_postgres_database.yml:
```diff
@@ -0,0 +1,14 @@
+# Database settings for postgres must be in sync with
+# test/dummy/config/database.yml.
```
Let's ✂️  this comment as well

adrianna-chang-shopify reviewed (approved):
Great work, thanks @shouichi ! A couple small comments. Let's also squash these commits before merging.

etiennebarrie commented on .github/workflows/ci.yml:
```diff
@@ -5,13 +5,27 @@ on: [push, pull_request]
       BUNDLE_GEMFILE: ${{ matrix.gemfile }}
+      DB: ${{ matrix.database }}
+    services:
+      postgres:
```
Is there a way to have a service run conditionally? Better yet, the environment that runs CI already has postgres, so we shouldn't need an external service like this:
https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#databases

shouichi replied:
> Is there a way to have a service run conditionally?

As far as I can tell, there's no way to start a service conditionally. https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idservices

The github guide (https://docs.github.com/en/actions/guides/creating-postgresql-service-containers) uses docker for postgres, so it shouldn't be that bad. Also, using containers gives us reproducibility and composability which may amortize the cost of spinning up extra service when sqlite testing.

etiennebarrie reviewed (commented):
> could I get thoughts / feedback on doing the above ^ to handle testing the install generator properly with PostgreSQL?

How about using `DATABASE_URL` which overrides `config/database.yml` entirely, just for the postgres cases in the matrix?

adrianna-chang-shopify commented:
> How about using DATABASE_URL which overrides config/database.yml entirely, just for the postgres cases in the matrix?

I think we'll still run into the same problem -- if we use a db with the same name (`maintenance_tasks_test`), Postgres will complain that the `maintenance_tasks_runs` table exists already in the install generator test. We could use a `DATABASE_URL` instead of the additional db config file, but I believe we'd need to have a different database. We could use psql to create one in CI, but this makes things not-so-nice for local testing. 

Also looked into starting PostgreSQL directly on the Ubuntu VM, but couldn't quite get it -- not sure how to provide a password and / or specify no password. I tried
```bash
  sudo systemctl start postgresql.service
  psql -h localhost -U postgres -w --no-password 'create database maintenance_tasks_test;'
---
  sudo systemctl start postgresql.service
  PG_PASSWORD=postgres psql -h localhost -U postgres 'create database maintenance_tasks_test;'
---
  sudo systemctl start postgresql.service
  sudo -u postgres psql postgres -c 'create database maintenance_tasks_test;'
```

But no luck with any of them. I think it's acceptable to start postgres in another service. If you'd like to invest some more time into it @etiennebarrie let me know -- otherwise I'll ship this tmrw afternoon.

shouichi commented:
Thanks for the feedback. Speaking of the local development, I plan to open a follow-up PR to introduce docker based development environment. This way we don't need to set up postgres locally. Does this make sense?

adrianna-chang-shopify commented:
Agree that `DATABASE_URL` is a more correct solution to this problem 👍  So to clarify @shouichi, let's do the following before shipping:
- Add a step to CI to export the `DATABASE_URL` if the db is postgres, as Étienne did in 3b8f99edfc1558b8eed3437245c67c877ccfb219 (although to me it's more readable to have it as `echo "DATABASE_URL=postgresql://localhost/maintenance_tasks_test" >> $GITHUB_ENV`)
- Remove the changes to `config/database.yml`
- Change the install generator to use the `DATABASE_URL` instead, as proposed in 631c89df519b7ace188f2350fa4a981e4b2c9863

Thanks for your patience! 🙏

shouichi commented:
@adrianna-chang-shopify Thanks for the clarification. The `DATABASE_URL` solution is much cleaner, it's great! Changed based on the suggestions, rebased, and squashed. Please take a look.
